### PR TITLE
chore(trie): simplify hashed cursor abstraction

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Builds on top of #8377.

Refactor hashed cursors abstractions. Unify common methods to a single `HashedCursor`. Make `HashedStorageCursor` trait extend the `HashedCursor` trait.

### CODE REVIEW

File: crates/trie/src/hashed_cursor/default.rs [hunk 1]
```diff
@@ -1,54 +1,85 @@
...
 }
 
-impl<C> HashedStorageCursor for C
+/// The structure wrapping a database cursor for hashed storage and
+/// a target hashed address. Implements [HashedCursor] and [HashedStorageCursor]
+/// for iterating over hashed storage.
+#[derive(Debug)]
+pub struct DatabaseHashedStorageCursor<C> {
+    cursor: C,
+    hashed_address: B256,
```
Comment (mattsse) [inline crates/trie/src/hashed_cursor/default.rs:50]: could we get one line docs for the fields, I'd fine this easier to understand, even though the wrapper concept is easy to follow.

File: crates/trie/src/hashed_cursor/post_state.rs [hunk 2]
```diff
@@ -179,36 +184,36 @@ pub struct HashedPostStateStorageCursor<'b, C> {
...
     post_state_storage_index: usize,
-    /// The current hashed account key.
-    account: Option<B256>,
     /// The last slot that has been returned by the cursor.
     /// De facto, this is the cursor's position for the given account key.
     last_slot: Option<B256>,
 }
 
 impl<'b, C> HashedPostStateStorageCursor<'b, C> {
     /// Create new instance of [HashedPostStateStorageCursor].
```
Comment (mattsse) [inline crates/trie/src/hashed_cursor/post_state.rs:197]: this should mention something like at the given hashed address

and should add a note that this does not check whether the storage is empty I think?

**Review by mattsse — APPROVED** [review]
I only have a few doc suggestions

File: crates/trie/src/hashed_cursor/post_state.rs [hunk 3]
```diff
@@ -219,62 +224,38 @@ impl<'b, C> HashedPostStateStorageCursor<'b, C> {
     /// If the storage keys are the same, the post state entry is given precedence.
     fn next_slot(
         post_state_item: Option<&(B256, U256)>,
-        db_item: Option<StorageEntry>,
-    ) -> Option<StorageEntry> {
+        db_item: Option<(B256, U256)>,
+    ) -> Option<(B256, U256)> {
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/trie/src/hashed_cursor/post_state.rs:228]: so, that `HashedCursor` can be generic over value

File: crates/trie/src/node_iter.rs [hunk 4]
```diff
@@ -236,15 +232,15 @@ where
             }
 
             // Attempt to get the next unprocessed key from the walker.
-            if let Some(seek_key) = self.walker.next_unprocessed_key() {
-                // Seek and update the current hashed entry based on the new seek key.
-                self.current_hashed_entry =
-                    self.hashed_storage_cursor.seek(self.hashed_address, seek_key)?;
-                self.walker.advance()?;
-            } else {
+            match self.walker.next_unprocessed_key() {
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/trie/src/node_iter.rs:235]: to resemble how it's done in `HashedAccountCursor`, it's dedupped later in anyway https://github.com/paradigmxyz/reth/pull/8381

**Pull request merged**